File: 0118-pascals-triangle/0118-pascals-triangle.go
/*
func generate(numRows int) [][]int {
    result := initializeSlice(numRows)
    for i := 1; i < numRows; i++ {
        for j := 1; j < i; j++ {
            result[i][j] = result[i-1][j-1] + result[i-1][j]
        }
    }
    
    return result
}

func initializeSlice(rows int) [][]int {
    s := make([][]int, rows)
    for i := 0; i < rows; i++ {
        for j := 0; j < i+1; j++ {
            s[i] = append(s[i], 1)
        }
    }
    return s
}
*/

func generate(numRows int) [][]int {
    result := [][]int{}
    above := []int{1}
    for i := 0; i < numRows; i++ {
        newRow := []int{}
        for j := 0; j <= i; j++ {
            if j == 0 || j == i {
                newRow = append(newRow, 1)
            } else {
                newRow = append(newRow, above[j] + above[j-1])
            }
        }
        result = append(result, newRow)
        above = newRow
    }
    return result
}

File: 0118-pascals-triangle/0118-pascals-triangle.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        if numRows == 0:
            return
        
        row1 = [1]
        result = [row1]
        above = row1
        for i in range(1, numRows):
            new_row = []
            for j in range(i+1):
                if j == 0 or j == i:
                    new_row.append(1)
                else:
                    new_row.append(above[j]+above[j-1])
                    
            result.append(new_row)
            above = new_row
            
        return result
        